### ISSUE TITLE
Literal promotion can lead to unsatisfiable solutions

### ISSUE THREAD

**Issue by ibraheemdev** [astral-sh/ty#4300]
`solution.promote(...)` replaces [the original solution] without checking whether the promoted type still satisfies the inferred upper bound. The existing check only considers an explicitly declared `TypeVar` bound.

For example:

```py
from typing import Callable, Literal

def make_list[T](value: T, accept: Callable[[T], None]) -> list[T]:
    return [value]

def accepts_one(value: Literal[1]) -> None:
    pass

reveal_type(make_list(1, accepts_one))
```

`T = Literal[1]` satisfies both arguments, so this call should succeed and reveal `list[Literal[1]]`. Instead, the PR reveals `list[int]` and incorrectly rejects `accepts_one`.

The same failure reproduces with a nested contravariant generic: passing `Consumer[Literal[1]]` to a parameter of type `Consumer[T]` incorrectly produces `Consumer[int]` and an `invalid-argument-type` diagnostic.

_Originally posted by @carljm in https://github.com/astral-sh/ruff/issues/27763#issuecomment-5298518302_
